File: frigo/dates.py
"""Analyse des dates de peremption saisies au pave numerique du terminal.

Le clavier du Skorpio est avant tout numerique : on accepte donc des saisies
courtes et sans separateur, ainsi que des raccourcis relatifs.

    ""            -> None (pas de date connue)
    "+7" / "7j"   -> aujourd'hui + 7 jours
    "2510"        -> 25 octobre (prochaine occurrence)
    "251026"      -> 25/10/2026
    "25102026"    -> 25/10/2026
    "1026"        -> ambigu : traite comme jour/mois (10 octobre)
    "10/2026"     -> 31/10/2026 (fin de mois, pour les DDM type "10/2026")
    "2026-10-25"  -> 25/10/2026
"""

from __future__ import annotations

import calendar
import datetime as dt
import re

__all__ = ["parse_expiry", "format_expiry", "days_left", "ExpiryError"]


class ExpiryError(ValueError):
    """Saisie de date impossible a interpreter."""


_REL = re.compile(r"^\+?(\d{1,4})\s*(?:j|jours?|d|days?)?$", re.I)
_ISO = re.compile(r"^(\d{4})[-/.](\d{1,2})[-/.](\d{1,2})$")
_DMY = re.compile(r"^(\d{1,2})[-/.](\d{1,2})[-/.](\d{2}|\d{4})$")
_MY = re.compile(r"^(\d{1,2})[-/.](\d{4})$")


def _end_of_month(year: int, month: int) -> dt.date:
    return dt.date(year, month, calendar.monthrange(year, month)[1])


def _expand_year(yy: int, today: dt.date) -> int:
    """Complete une annee sur 2 chiffres dans une fenetre -10/+89 ans."""
    century = today.year - today.year % 100
    year = century + yy
    if year < today.year - 10:
        year += 100
    return year


def parse_expiry(text: str | None, today: dt.date | None = None) -> dt.date | None:
    """Convertit une saisie terminal en date. Retourne None si vide."""
    today = today or dt.date.today()
    if text is None:
        return None
    raw = text.strip()
    if not raw or raw in {"-", "0"}:
        return None

    # Raccourci relatif : "+7", "7j", "30"...  (uniquement si prefixe +, ou <= 3
    # chiffres qui ne ressemblent pas a une date compacte)
    m = _REL.match(raw)
    if m and (raw.startswith("+") or len(m.group(1)) <= 3):
        return today + dt.timedelta(days=int(m.group(1)))

    if m := _ISO.match(raw):
        y, mo, d = (int(g) for g in m.groups())
        return _safe_date(y, mo, d)

    if m := _DMY.match(raw):
        d, mo, y = (int(g) for g in m.groups())
        if len(m.group(3)) == 2:
            y = _expand_year(y, today)
        return _safe_date(y, mo, d)

    if m := _MY.match(raw):
        mo, y = int(m.group(1)), int(m.group(2))
        if not 1 <= mo <= 12:
            raise ExpiryError(f"mois invalide dans {raw!r}")
        return _end_of_month(y, mo)

    digits = re.sub(r"\D", "", raw)
    if digits != raw and not digits:
        raise ExpiryError(f"date illisible : {raw!r}")

    if len(digits) == 4:  # DDMM
        d, mo = int(digits[:2]), int(digits[2:])
        if not 1 <= mo <= 12:
            # peut-etre MMYY saisi a l'envers -> fin de mois
            mo2, yy = int(digits[:2]), int(digits[2:])
            if 1 <= mo2 <= 12:
                return _end_of_month(_expand_year(yy, today), mo2)
            raise ExpiryError(f"date illisible : {raw!r}")
        candidate = _safe_date(today.year, mo, d)
        if candidate < today:  # une DDM est presque toujours dans le futur
            candidate = _safe_date(today.year + 1, mo, d)
        return candidate

    if len(digits) == 6:  # DDMMYY
        d, mo, yy = int(digits[:2]), int(digits[2:4]), int(digits[4:])
        return _safe_date(_expand_year(yy, today), mo, d)

    if len(digits) == 8:  # DDMMYYYY
        d, mo, y = int(digits[:2]), int(digits[2:4]), int(digits[4:])
        return _safe_date(y, mo, d)

    raise ExpiryError(f"date illisible : {raw!r}")


def _safe_date(year: int, month: int, day: int) -> dt.date:
    try:
        return dt.date(year, month, day)
    except ValueError as exc:
        raise ExpiryError(f"date inexistante : {day:02d}/{month:02d}/{year}") from exc


def format_expiry(value: dt.date | str | None) -> str:
    """Affichage court pour le terminal (8 caracteres)."""
    if value is None:
        return "-"
    if isinstance(value, str):
        try:
            value = dt.date.fromisoformat(value)
        except ValueError:
            return value
    return value.strftime("%d/%m/%y")


def days_left(value: dt.date | str | None, today: dt.date | None = None) -> int | None:
    """Nombre de jours restants (negatif si perime)."""
    if value is None:
        return None
    if isinstance(value, str):
        try:
            value = dt.date.fromisoformat(value)
        except ValueError:
            return None
    return (value - (today or dt.date.today())).days

File: client/SkorpioFrigo/Dates.cs
using System;

namespace SkorpioFrigo
{
    /// <summary>
    /// Saisie de date au pave numerique, miroir de frigo/dates.py.
    ///
    /// Le Skorpio-G a un pave de 28 ou 38 touches : pas de '/' commode, pas de
    /// selecteur de date. On accepte donc la frappe la plus courte possible.
    ///
    ///     "2510"      25 octobre (prochaine occurrence)
    ///     "251026"    25/10/2026
    ///     "25102026"  25/10/2026
    ///     ""          date inconnue
    ///
    /// En mode relatif (touche J+), le nombre saisi est un nombre de jours :
    /// "7" devient la date d'aujourd'hui plus sept jours.
    /// </summary>
    internal static class Dates
    {
        /// <summary>
        /// Analyse une saisie. Retourne true si la saisie est exploitable ;
        /// result vaut DateTime.MinValue quand l'utilisateur laisse le champ vide.
        /// </summary>
        public static bool TryParse(string text, bool relative, out DateTime result,
                                    out string error)
        {
            result = DateTime.MinValue;
            error = null;
            string digits = OnlyDigits(text);

            if (digits.Length == 0)
                return true;                            // date volontairement inconnue

            DateTime today = DateTime.Now.Date;

            if (relative)
            {
                int days;
                if (!TryInt(digits, out days) || days < 0 || days > 3650)
                {
                    error = "nombre de jours invalide";
                    return false;
                }
                result = today.AddDays(days);
                return true;
            }

            int d, m, y;
            switch (digits.Length)
            {
                case 4:                                 // JJMM
                    if (!TryInt(digits.Substring(0, 2), out d) ||
                        !TryInt(digits.Substring(2, 2), out m))
                    {
                        error = "date illisible";
                        return false;
                    }
                    y = today.Year;
                    if (!Build(y, m, d, out result, out error))
                        return false;
                    if (result < today)                 // une DDM est dans le futur
                        return Build(y + 1, m, d, out result, out error);
                    return true;

                case 6:                                 // JJMMAA
                    if (!TryInt(digits.Substring(0, 2), out d) ||
                        !TryInt(digits.Substring(2, 2), out m) ||
                        !TryInt(digits.Substring(4, 2), out y))
                    {
                        error = "date illisible";
                        return false;
                    }
                    return Build(ExpandYear(y, today), m, d, out result, out error);

                case 8:                                 // JJMMAAAA
                    if (!TryInt(digits.Substring(0, 2), out d) ||
                        !TryInt(digits.Substring(2, 2), out m) ||
                        !TryInt(digits.Substring(4, 4), out y))
                    {
                        error = "date illisible";
                        return false;
                    }
                    return Build(y, m, d, out result, out error);

                default:
                    error = "attendu JJMM, JJMMAA ou JJMMAAAA";
                    return false;
            }
        }

        /// <summary>Complete une annee sur deux chiffres dans une fenetre -10/+89 ans.</summary>
        private static int ExpandYear(int yy, DateTime today)
        {
            int century = today.Year - (today.Year % 100);
            int year = century + yy;
            if (year < today.Year - 10)
                year += 100;
            return year;
        }

        private static bool Build(int y, int m, int d, out DateTime result, out string error)
        {
            result = DateTime.MinValue;
            error = null;
            if (m < 1 || m > 12)
            {
                error = "mois " + m + " invalide";
                return false;
            }
            if (y < 1970 || y > 2999)
            {
                error = "annee " + y + " invalide";
                return false;
            }
            if (d < 1 || d > DateTime.DaysInMonth(y, m))
            {
                error = "jour " + d + " invalide pour ce mois";
                return false;
            }
            result = new DateTime(y, m, d);
            return true;
        }

        public static string ToIso(DateTime value)
        {
            return value == DateTime.MinValue ? "" : value.ToString("yyyy-MM-dd");
        }

        public static string ToShort(DateTime value)
        {
            return value == DateTime.MinValue ? "sans date" : value.ToString("dd/MM/yy");
        }

        /// <summary>Jours restants, pour colorer l'ecran apres un scan.</summary>
        public static int DaysLeft(DateTime value)
        {
            if (value == DateTime.MinValue)
                return int.MaxValue;
            return (int)(value.Date - DateTime.Now.Date).TotalDays;
        }

        private static string OnlyDigits(string text)
        {
            if (text == null)
                return "";
            char[] buf = new char[text.Length];
            int n = 0;
            for (int i = 0; i < text.Length; i++)
            {
                if (text[i] >= '0' && text[i] <= '9')
                    buf[n++] = text[i];
            }
            return new string(buf, 0, n);
        }

        // int.TryParse existe sous CF 2.0 mais reste peu fiable selon les images :
        // une conversion explicite evite toute surprise a l'execution.
        private static bool TryInt(string text, out int value)
        {
            value = 0;
            try
            {
                value = int.Parse(text);
                return true;
            }
            catch (Exception)
            {
                return false;
            }
        }
    }
}
